File: src/preprocessing/post.ts
import type {
  Post,
  ProcessedAttachment,
  ProcessedFile,
  ProcessedMessage,
  ReactionActor,
  ReactionSummary,
  Redaction,
  User,
} from '../types'
import { setCanonicalPostIdentity } from '../utils/threading'
import { preprocess } from './pipeline'

const DELETED_POST_TEXT = '[deleted post]'
type UserLookup = ReadonlyMap<string, User>

function record(value: unknown): Record<string, unknown> | undefined {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
    ? (value as Record<string, unknown>)
    : undefined
}

function stringValue(value: unknown, allowNumber = false): string | undefined {
  if (typeof value === 'string') return value
  if (allowNumber && typeof value === 'number' && Number.isFinite(value)) return String(value)
  return undefined
}

function arrayValue(value: unknown): unknown[] {
  return Array.isArray(value) ? value : []
}

export function postUserIds(posts: Post[]): string[] {
  const ids = new Set<string>()
  for (const post of posts) {
    if (post.user_id) ids.add(post.user_id)
    for (const candidate of arrayValue(post.metadata?.reactions)) {
      const id = stringValue(record(candidate)?.user_id)
      if (id) ids.add(id)
    }
  }
  return [...ids]
}

export function normalizePosts(
  posts: Post[],
  users: UserLookup,
  myUserId: string,
  serverUrl: string,
  buildPermalink: (serverUrl: string, postId: string) => string,
  redact: boolean,
): { messages: ProcessedMessage[]; redactions: Redaction[] } {
  const allRedactions: Redaction[] = []
  const clean = (value: string, field: string, oneLine = false): string => {
    const result = preprocess(value, { redact })
    allRedactions.push(...result.redactions.map((item) => ({ ...item, field })))
    return oneLine ? result.text.replace(/\n/g, '\\n').replace(/\t/g, '\\t') : result.text
  }

  const messages = posts.map((post): ProcessedMessage => {
    const isDeleted = post.delete_at > 0
    const props = record(post.props) ?? {}
    const rawOverride = stringValue(post.override_username) ?? stringValue(props.override_username)
    const rawUsername = users.get(post.user_id)?.username ?? post.user_id
    const displayUser = rawOverride
      ? clean(rawOverride, 'user', true)
      : !post.user_id || post.type.startsWith('system_')
        ? 'system'
        : post.user_id === myUserId
          ? 'you'
          : clean(rawUsername, 'user', true)

    const rawFiles = arrayValue(post.metadata?.files).flatMap((candidate) => {
      const value = record(candidate)
      const id = stringValue(value?.id)
      return value && id ? [{ value, id }] : []
    })
    const metadataFiles = new Map(rawFiles.map(({ id, value }) => [id, value]))
    const rawFileIds = [
      ...new Set([
        ...arrayValue(post.file_ids).flatMap((value) => {
          const id = stringValue(value)
          return id ? [id] : []
        }),
        ...metadataFiles.keys(),
      ]),
    ]
    const fileDetails: ProcessedFile[] = isDeleted
      ? []
      : rawFileIds.map((rawId) => {
          const file = metadataFiles.get(rawId)
          const name = stringValue(file?.name)
          const mime = stringValue(file?.mime_type)
          const extension = stringValue(file?.extension)
          return {
            id: clean(rawId, 'file.id', true),
            ...(name ? { name: clean(name, 'file.name', true) } : {}),
            ...(mime ? { mime: clean(mime, 'file.mime', true) } : {}),
            ...(typeof file?.size === 'number' && Number.isFinite(file.size)
              ? { size: file.size }
              : {}),
            ...(extension ? { extension: clean(extension, 'file.extension', true) } : {}),
          }
        })

    const attachments: ProcessedAttachment[] = isDeleted
      ? []
      : arrayValue(props.attachments).flatMap((candidate, index) => {
          const source = record(candidate)
          if (!source) return []
          const cleanedValues = new Map<string, string | undefined>()
          const take = (key: string, oneLine = false, allowNumber = false) => {
            const cacheKey = `${key}:${oneLine}:${allowNumber}`
            if (cleanedValues.has(cacheKey)) return cleanedValues.get(cacheKey)
            const value = stringValue(source[key], allowNumber)
            const cleaned = value ? clean(value, `attachment.${index}.${key}`, oneLine) : undefined
            cleanedValues.set(cacheKey, cleaned)
            return cleaned
          }
          const fields = arrayValue(source.fields).flatMap((candidateField, fieldIndex) => {
            const field = record(candidateField)
            if (!field) return []
            const title = stringValue(field.title, true)
            const value = stringValue(field.value, true)
            if (!title && !value && typeof field.short !== 'boolean') return []
            return [
              {
                ...(title
                  ? { title: clean(title, `attachment.${index}.fields.${fieldIndex}.title`) }
                  : {}),
                ...(value
                  ? { value: clean(value, `attachment.${index}.fields.${fieldIndex}.value`) }
                  : {}),
                ...(typeof field.short === 'boolean' ? { short: field.short } : {}),
              },
            ]
          })
          const attachment: ProcessedAttachment = {
            ...(take('fallback') ? { fallback: take('fallback') } : {}),
            ...(take('pretext') ? { pretext: take('pretext') } : {}),
            ...(take('title') ? { title: take('title') } : {}),
            ...(take('title_link', true) ? { titleLink: take('title_link', true) } : {}),
            ...(take('text') ? { text: take('text') } : {}),
            ...(fields.length > 0 ? { fields } : {}),
            ...(take('footer') ? { footer: take('footer') } : {}),
            ...(take('footer_icon', true) ? { footerIcon: take('footer_icon', true) } : {}),
            ...(take('author_name') ? { authorName: take('author_name') } : {}),
            ...(take('author_link', true) ? { authorLink: take('author_link', true) } : {}),
            ...(take('author_icon', true) ? { authorIcon: take('author_icon', true) } : {}),
            ...(take('color', true) ? { color: take('color', true) } : {}),
            ...(take('image_url', true) ? { imageUrl: take('image_url', true) } : {}),
            ...(take('thumb_url', true) ? { thumbUrl: take('thumb_url', true) } : {}),
            ...(take('ts', true, true) ? { timestamp: take('ts', true, true) } : {}),
          }
          return Object.keys(attachment).length > 0 ? [attachment] : []
        })

    const reactionGroups = new Map<string, Array<{ rawId: string; actor: ReactionActor }>>()
    if (!isDeleted) {
      for (const candidate of arrayValue(post.metadata?.reactions)) {
        const reaction = record(candidate)
        const rawEmoji = stringValue(reaction?.emoji_name)
        const rawId = stringValue(reaction?.user_id)
        if (!rawEmoji || !rawId) continue
        const username = users.get(rawId)?.username
        const actor: ReactionActor = {
          id: clean(rawId, 'reaction.actor.id', true),
          ...(username ? { username: clean(username, 'reaction.actor.username', true) } : {}),
        }
        const group = reactionGroups.get(rawEmoji)
        if (group) group.push({ rawId, actor })
        else reactionGroups.set(rawEmoji, [{ rawId, actor }])
      }
    }
    const reactions: ReactionSummary[] = [...reactionGroups]
      .sort(([a], [b]) => a.localeCompare(b))
      .map(([rawEmoji, entries]) => {
        entries.sort((a, b) => a.rawId.localeCompare(b.rawId))
        return {
          emoji: clean(rawEmoji, 'reaction.emoji', true),
          count: entries.length,
          actors: entries.map(({ actor }) => actor),
        }
      })

    const message: ProcessedMessage = {
      id: clean(post.id, 'post.id', true),
      permalink: clean(buildPermalink(serverUrl, post.id), 'post.permalink', true),
      user: displayUser,
      userId: clean(post.user_id, 'post.userId', true),
      text: isDeleted ? DELETED_POST_TEXT : clean(post.message, 'post.message'),
      timestamp: new Date(post.create_at),
      updatedAt: new Date(post.update_at),
      ...(post.edit_at > 0 ? { editedAt: new Date(post.edit_at) } : {}),
      ...(post.delete_at > 0 ? { deletedAt: new Date(post.delete_at) } : {}),
      isDeleted,
      postType: clean(post.type, 'post.type', true),
      isSystem: post.type.startsWith('system_') || !post.user_id,
      isPinned: post.is_pinned ?? false,
      files: isDeleted ? [] : rawFileIds.map((id) => clean(id, 'file.id', true)),
      fileDetails,
      attachments,
      reactions,
      rootId: post.root_id ? clean(post.root_id, 'post.rootId', true) : undefined,
      replyCount: post.reply_count || undefined,
    }
    setCanonicalPostIdentity(message, post.id, post.root_id || undefined)
    return message
  })
  return { messages, redactions: allRedactions }
}

File: internal/normalization/post.go
// Package normalization converts validated Mattermost data into safe output models.
package normalization

import (
	"sort"
	"strings"
	"time"
	"unicode/utf16"

	"github.com/ardasevinc/mattermost-cli/internal/mattermost"
	"github.com/ardasevinc/mattermost-cli/internal/output"
	"github.com/ardasevinc/mattermost-cli/internal/presentation"
	"github.com/ardasevinc/mattermost-cli/internal/serverurl"
)

const deletedPostText = "[deleted post]"

// PostUserIDs returns unique author and reaction actor IDs in encounter order.
func PostUserIDs(posts []mattermost.Post) []string {
	seen := make(map[string]struct{})
	ids := make([]string, 0)
	add := func(id string) {
		if id == "" {
			return
		}
		if _, ok := seen[id]; ok {
			return
		}
		seen[id] = struct{}{}
		ids = append(ids, id)
	}
	for _, post := range posts {
		add(post.UserID)
		for _, reaction := range post.Reactions {
			add(reaction.UserID)
		}
	}
	return ids
}

// NormalizePosts safely prepares validated posts for presentation.
func NormalizePosts(posts []mattermost.Post, users map[string]mattermost.User, myUserID, serverURL string, options presentation.Options) ([]output.Message, []output.Redaction, error) {
	redactions := make([]output.Redaction, 0)
	clean := func(value, field string, label bool) string {
		result := presentation.PreprocessWithOptions(value, options)
		if label {
			remapLabelRedactionPositions(result.Text, result.Redactions)
			result.Text = presentation.SanitizeLabel(result.Text)
		}
		for _, item := range result.Redactions {
			item.Field = field
			redactions = append(redactions, item)
		}
		return result.Text
	}
	messages := make([]output.Message, 0, len(posts))
	for _, post := range posts {
		permalink, err := serverurl.BuildPostPermalink(serverURL, post.ID)
		if err != nil {
			return nil, nil, err
		}
		deleted := post.DeleteAt > 0
		username := post.UserID
		if user, ok := users[post.UserID]; ok {
			username = user.Username
		}
		displayUser := ""
		switch {
		case post.OverrideUsername != "":
			displayUser = clean(post.OverrideUsername, "user", true)
		case post.UserID == "" || strings.HasPrefix(post.Type, "system_"):
			displayUser = "system"
		case post.UserID == myUserID:
			displayUser = "you"
		default:
			displayUser = clean(username, "user", true)
		}

		files, details := normalizeFiles(post, deleted, clean)
		attachments := normalizeAttachments(post, deleted, clean)
		reactions := normalizeReactions(post, users, deleted, clean)
		messageText := deletedPostText
		if !deleted {
			messageText = clean(post.Message, "post.message", false)
		}
		message := output.Message{
			ID: clean(post.ID, "post.id", true), Permalink: clean(permalink, "post.permalink", true),
			User: displayUser, UserID: clean(post.UserID, "post.userId", true), Text: messageText,
			Timestamp: time.UnixMilli(post.CreateAt).UTC(), UpdatedAt: time.UnixMilli(post.UpdateAt).UTC(),
			IsDeleted: deleted, PostType: clean(post.Type, "post.type", true),
			IsSystem: post.UserID == "" || strings.HasPrefix(post.Type, "system_"), IsPinned: post.IsPinned,
			Files: files, FileDetails: details, Attachments: attachments, Reactions: reactions,
			CanonicalID: post.ID, CanonicalRootID: post.RootID,
		}
		shapeKnown := post.ThreadShapeKnown
		message.CanonicalThreadShapeKnown = &shapeKnown
		if post.EditAt > 0 {
			value := time.UnixMilli(post.EditAt).UTC()
			message.EditedAt = &value
		}
		if post.DeleteAt > 0 {
			value := time.UnixMilli(post.DeleteAt).UTC()
			message.DeletedAt = &value
		}
		if post.RootID != "" {
			message.RootID = clean(post.RootID, "post.rootId", true)
		}
		if post.ReplyCount > 0 {
			value := post.ReplyCount
			message.ReplyCount = &value
		}
		messages = append(messages, message)
	}
	return messages, redactions, nil
}

type cleaner func(string, string, bool) string

func normalizeFiles(post mattermost.Post, deleted bool, clean cleaner) ([]string, []output.File) {
	if deleted {
		return []string{}, []output.File{}
	}
	metadata := make(map[string]mattermost.PostFile, len(post.Files))
	for _, file := range post.Files {
		if file.ID != "" {
			metadata[file.ID] = file
		}
	}
	ids := make([]string, 0, len(post.FileIDs)+len(post.Files))
	seen := make(map[string]struct{})
	add := func(id string) {
		if id != "" {
			if _, ok := seen[id]; !ok {
				seen[id] = struct{}{}
				ids = append(ids, id)
			}
		}
	}
	for _, id := range post.FileIDs {
		add(id)
	}
	for _, file := range post.Files {
		add(file.ID)
	}
	visible := make([]string, 0, len(ids))
	details := make([]output.File, 0, len(ids))
	for _, id := range ids {
		visible = append(visible, clean(id, "file.id", true))
		source := metadata[id]
		item := output.File{ID: clean(id, "file.id", true)}
		if source.Name != "" {
			item.Name = clean(source.Name, "file.name", true)
		}
		if source.MIMEType != "" {
			item.MIME = clean(source.MIMEType, "file.mime", true)
		}
		if source.Size != nil {
			value := *source.Size
			item.Size = &value
		}
		if source.Extension != "" {
			item.Extension = clean(source.Extension, "file.extension", true)
		}
		details = append(details, item)
	}
	return visible, details
}

func normalizeAttachments(post mattermost.Post, deleted bool, clean cleaner) []output.Attachment {
	result := make([]output.Attachment, 0)
	if deleted {
		return result
	}
	for i, source := range post.Attachments {
		field := func(name string) string { return "attachment." + itoa(i) + "." + name }
		item := output.Attachment{
			Fallback: cleanOptional(source.Fallback, field("fallback"), false, clean), Pretext: cleanOptional(source.Pretext, field("pretext"), false, clean),
			Title: cleanOptional(source.Title, field("title"), false, clean), TitleLink: cleanOptional(source.TitleLink, field("title_link"), true, clean),
			Text: cleanOptional(source.Text, field("text"), false, clean), Footer: cleanOptional(source.Footer, field("footer"), false, clean),
			FooterIcon: cleanOptional(source.FooterIcon, field("footer_icon"), true, clean), AuthorName: cleanOptional(source.AuthorName, field("author_name"), false, clean),
			AuthorLink: cleanOptional(source.AuthorLink, field("author_link"), true, clean), AuthorIcon: cleanOptional(source.AuthorIcon, field("author_icon"), true, clean),
			Color: cleanOptional(source.Color, field("color"), true, clean), ImageURL: cleanOptional(source.ImageURL, field("image_url"), true, clean),
			ThumbURL: cleanOptional(source.ThumbURL, field("thumb_url"), true, clean), Timestamp: cleanOptional(source.Timestamp, field("ts"), true, clean),
		}
		for j, sourceField := range source.Fields {
			prefix := field("fields." + itoa(j))
			f := output.AttachmentField{Title: cleanOptional(sourceField.Title, prefix+".title", false, clean), Value: cleanOptional(sourceField.Value, prefix+".value", false, clean)}
			if sourceField.Short != nil {
				value := *sourceField.Short
				f.Short = &value
			}
			if f.Title != "" || f.Value != "" || f.Short != nil {
				item.Fields = append(item.Fields, f)
			}
		}
		result = append(result, item)
	}
	return result
}

func normalizeReactions(post mattermost.Post, users map[string]mattermost.User, deleted bool, clean cleaner) []output.Reaction {
	result := make([]output.Reaction, 0)
	if deleted {
		return result
	}
	groups := make(map[string][]mattermost.PostReaction)
	for _, reaction := range post.Reactions {
		if reaction.EmojiName != "" && reaction.UserID != "" {
			groups[reaction.EmojiName] = append(groups[reaction.EmojiName], reaction)
		}
	}
	emojis := make([]string, 0, len(groups))
	for emoji := range groups {
		emojis = append(emojis, emoji)
	}
	sort.Strings(emojis)
	for _, emoji := range emojis {
		entries := groups[emoji]
		sort.SliceStable(entries, func(i, j int) bool { return entries[i].UserID < entries[j].UserID })
		actors := make([]output.ReactionActor, 0, len(entries))
		for _, entry := range entries {
			actor := output.ReactionActor{ID: clean(entry.UserID, "reaction.actor.id", true)}
			if user, ok := users[entry.UserID]; ok && user.Username != "" {
				actor.Username = clean(user.Username, "reaction.actor.username", true)
			}
			actors = append(actors, actor)
		}
		result = append(result, output.Reaction{Emoji: clean(emoji, "reaction.emoji", true), Count: len(entries), Actors: actors})
	}
	return result
}

func cleanOptional(value, field string, label bool, clean cleaner) string {
	if value == "" {
		return ""
	}
	return clean(value, field, label)
}

func remapLabelRedactionPositions(text string, redactions []presentation.Redaction) {
	next := 0
	originalPosition := 0
	expansion := 0
	applyThrough := func(position int) {
		for next < len(redactions) && redactions[next].Position <= position {
			redactions[next].Position += expansion
			next++
		}
	}
	applyThrough(0)
	for _, character := range text {
		width := 1
		if character > 0xffff {
			width = 2
		}
		originalPosition += width
		if character == '\n' || character == '\t' {
			expansion++
		}
		applyThrough(originalPosition)
	}
	for next < len(redactions) {
		redactions[next].Position += expansion
		next++
	}
}

func utf16Length(value string) int { return len(utf16.Encode([]rune(value))) }

func itoa(value int) string {
	if value == 0 {
		return "0"
	}
	var digits [20]byte
	cursor := len(digits)
	for value > 0 {
		cursor--
		digits[cursor] = byte('0' + value%10)
		value /= 10
	}
	return string(digits[cursor:])
}
